pkg/provider/cloud/azure/availability_set.go

```diff
@@ -0,0 +1,119 @@
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package azure
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/Azure/azure-sdk-for-go/services/compute/mgmt/2021-07-01/compute"
+	"github.com/Azure/azure-sdk-for-go/services/compute/mgmt/2021-07-01/compute/computeapi"
+	"github.com/Azure/go-autorest/autorest/to"
+
+	kubermaticv1 "k8c.io/kubermatic/v2/pkg/crd/kubermatic/v1"
+	kuberneteshelper "k8c.io/kubermatic/v2/pkg/kubernetes"
+	"k8c.io/kubermatic/v2/pkg/provider"
+)
+
+func availabilitySetName(cluster *kubermaticv1.Cluster) string {
+	return resourceNamePrefix + cluster.Name
+}
+
+func reconcileAvailabilitySet(ctx context.Context, clients *ClientSet, location string, cluster *kubermaticv1.Cluster, update provider.ClusterUpdater) (*kubermaticv1.Cluster, error) {
+	if cluster.Spec.Cloud.Azure.AvailabilitySet == "" {
+		cluster.Spec.Cloud.Azure.AvailabilitySet = availabilitySetName(cluster)
+	}
+
+	availabilitySet, err := clients.AvailabilitySets.Get(ctx, cluster.Spec.Cloud.Azure.ResourceGroup, cluster.Spec.Cloud.Azure.AvailabilitySet)
+	if err != nil && !isNotFound(availabilitySet.Response) {
+		return cluster, err
+	}
+
+	// if we found an availability set, we can check for the ownership tag to determine
+	// if the referenced availability set is owned by this cluster and should be reconciled
+	if !isNotFound(availabilitySet.Response) && !hasOwnershipTag(availabilitySet.Tags, cluster) {
+		return cluster, nil
+	}
+
+	target, err := targetAvailabilitySet(cluster.Spec.Cloud, location, cluster.Name)
+	if err != nil {
+		return cluster, err
+	}
+
+	// check for attributes of the existing availability set and return early if all values are already
+	// as expected. Since there are a lot of pointers in the compute.AvailabilitySet struct, we need to
+	// do a lot of "!= nil" checks so this does not panic.
+	//
+	// Attributes we check:
+	// - SKU name
+	// - fault domain count
+	// - update domain count
+	if (availabilitySet.Sku != nil && availabilitySet.Sku.Name != nil && *availabilitySet.Sku.Name == *target.Sku.Name) && availabilitySet.AvailabilitySetProperties != nil &&
+		(availabilitySet.AvailabilitySetProperties.PlatformFaultDomainCount != nil && *availabilitySet.AvailabilitySetProperties.PlatformFaultDomainCount == *target.AvailabilitySetProperties.PlatformFaultDomainCount) &&
+		(availabilitySet.AvailabilitySetProperties.PlatformUpdateDomainCount != nil && *availabilitySet.AvailabilitySetProperties.PlatformUpdateDomainCount == *target.AvailabilitySetProperties.PlatformUpdateDomainCount) {
+		return cluster, nil
+	}
```

Comment:
@moadqassem do you have any thoughts on this? I'm not happy with it, but due to the fact that Azure heavily relies on pointers, I can't do `reflect.DeepEqual` on the structs, at least that's my understanding.